openage/convert/dataformat/data_definition.py

```diff
@@ -124,6 +124,114 @@ def generate_csv(self, genfile):
+        txt.extend([
+            "# Image file path", "\n",
+            "image 0 ", "./", self.name_data_file, ".png", "\n",  # TODO: this sucks
+            "\n",
+        ])
+
+        # layer information
+        txt.extend([
+            "# Missing info about layers, bottom to top.", "\n",
+            "layer <layer_id> ...", "\n",
+            "\n",
+            ])
+
+        # for each angle, list the frame information
+        txt.extend([
+            "# Angle information comes from empires.dat. E.g.:", "\n",
+            "angle 270 mirror_from=90", "\n",
+            "\n",
+        ])
+
+        # actual frame information. TODO: Needs to be split based on frame angle
+        txt.extend([
+            "# Frame information:", "\n",
+            "# frame <layer_id> <image_id> <xpos> <ypos> <xsize> <ysize> <xhotspot> <yhotspot>", "\n",
+        ])
+
+        for idx, data_line in enumerate(self.data):
+            row_entries = list()
+            for member_name, member_type in self.members.items():
+                entry = data_line[member_name]
+
+                make_relpath = False
+
+                # check if enum data value is valid
+                if isinstance(member_type, EnumMember) and \
+                        not member_type.validate_value(entry):
+
+                    raise Exception("data entry %d '%s'"
+                                    " not a valid %s value" %
+                                    (idx, entry, repr(member_type)))
+
+                # insert filename to read this field
+                if isinstance(member_type, MultisubtypeMember):
+                    # subdata member stores the follow-up filename
+                    entry += GeneratedFile.output_preferences["sprite"]["file_suffix"]
+                    make_relpath = True
+
+                from .multisubtype_base import MultisubtypeBaseFile
+                if self.target == MultisubtypeBaseFile:
+                    # if the struct definition target is the multisubtype
+                    # base file, it already created the filename entry.
+                    # it needs to be made relative as well.
+                    if member_name == MultisubtypeBaseFile.data_format[1][1]:
+                        # only make the filename entry relative
+                        make_relpath = True
+
+                if make_relpath:
+                    # filename to reference to, make it relative to the
+                    # current file name
+                    entry = os.path.relpath(
+                        entry,
+                        os.path.dirname(self.name_data_file)
+                    ).replace(os.path.sep, '/')  # HACK: Change to better path handling
+
+                # encode each data field, to escape newlines and commas
+                row_entries.append(encode_value(entry))
+
+            # create one "frame" line with data separated by tab
+            txt.extend(("frame\t<layer_id>\t0\t", "\t".join(row_entries), "\n"))
```

Comment:
the layer ID (and image ID) is probably decided beforehand and should be part of the joined entries.

Reply:
We need to solve the above first.